Sources/Datadog/DatadogCore/Storage/Writing/FileWriter.swift

```diff
@@ -20,13 +20,30 @@ internal struct FileWriter: Writer {
 
     // MARK: - Writing data
 
-    /// Encodes given value to JSON data and writes it to the file.
-    func write<T: Encodable>(value: T) {
+    /// Encodes given encodable value and metadata, and writes it to the file.
+    /// If encryption is available, the data is encrypted before writing.
+    /// - Parameters:
+    ///  - value: Encodable value to write.
+    ///  - metadata: Encodable metadata to write.
+    func write<T: Encodable, M: Encodable>(value: T, metadata: M?) {
         do {
-            let data = try encode(event: value)
-            let writeSize = UInt64(data.count)
+            let encodedValue = try encode(encodable: value, blockType: .event)
+            let encodedMetadata: Data?
+            if let metadata = metadata {
+                encodedMetadata = try encode(encodable: metadata, blockType: .eventMetadata)
+            } else {
+                encodedMetadata = nil
+            }
+
+            // Make sure both event and event metadata are written to the same file.
+            // This is to avoid a situation where event is written to one file and event metadata to another.
+            // If this happens, the reader will not be able to match event with its metadata.
+            let writeSize = UInt64(encodedValue.count + (encodedMetadata?.count ?? 0))
             let file = try forceNewFile ? orchestrator.getNewWritableFile(writeSize: writeSize) : orchestrator.getWritableFile(writeSize: writeSize)
-            try file.append(data: data)
+            try file.append(data: encodedValue)
+            if let encodedMetadata = encodedMetadata {
+                try file.append(data: encodedMetadata)
+            }
```

Comment:
blocker/ Writes must be atomic (one transaction; both `event + meta` or nothing). Lack of atomicity was a severe issue in the past (incident-4797), so we'll run into regression with this code. 
* See: https://github.com/DataDog/dd-sdk-ios/pull/133
* Also see the number of I/O errors in iOS telemetry to get the idea on reliability of file writing 😉 

What we also need is making sure that `event + metadata` write is stress tested against flaky I/O. This test corresponds to incident-4797 and we can duplicate it / adapt to the new write variant:

https://github.com/DataDog/dd-sdk-ios/blob/73e99dd1d5813624bff9af5cb9196c3b7c6b9cf4/Tests/DatadogTests/Datadog/Core/Persistence/Writing/FileWriterTests.swift#L168-L169

🤔💭 Looking at this test, I think it became outdated after introduction of TLV during V2 migration (in V1 we were writing `data + separator byte`). One more argument to touch it. WDYT @ganeshnj ?

Reply:
This was great point, I couldn't able to fail the test with randomization we have placed and partially I think because we interrupt the writing after in between write calls. 

But like we talked, this test still has value so I modified with taking metadata into consideration.